### PR TITLE
Search Commits via Commit Hash

### LINKED ISSUE #7395: Search commit via commit SHA(short or full)
Currently it will always search in commit message.

### PR DESCRIPTION
**Pull request by gary-kim**
Closes #7395 
Commit history can now be searched with commit hash.

EDIT: whoops, slight mistake in the test case. I'll fix it in a bit.
EDIT2: should be fixed now.

### CODE REVIEW

File: modules/git/repo_commit.go
Comment (lunny) [inline modules/git/repo_commit.go]: I don't think this is right, because there is a prameter `-100` before.

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
Maybe we could add an extra `git log <keyword> -1` to get all output when `keyword` didn't containers `:` and then merge the output with original.

**Comment by gary-kim** [comment]
> Maybe we could add an extra `git log <keyword> -1` to get all output when `keyword` didn't containers `:` and then merge the output with original.

Wouldn't that only match a single commit hash? In Gitea's commit history, there are 4 hashes that start with `46d3` so doing `git log 46d3 -1` only outputs the most recent one. Also, do we want this matching only from the start or from the middle of the commit hash as well?

File: modules/git/repo_commit.go
Comment (gary-kim) [reply modules/git/repo_commit.go]: I didn't think about that. The commit hash search would not go back as far as the grep search.

**Comment by lunny** [comment]
@gary-kim I think we should only match prefix what is the `git log` does.

**Comment by gary-kim** [comment]
> @gary-kim I think we should only match prefix what is the `git log` does.

@lunny Okay, done.

File: integrations/repo_commits_search_test.go
Comment (lunny) [inline integrations/repo_commits_search_test.go:33]: Could we have a failed test.

**Comment by lunny** [comment]
@gary-kim Good work! Two small things we may need to consider:
- if a commit has another commit's id on commit message, we may get duplicated commit output, how to fix that?
- I find repo. parsePrettyFormatLogToList will get all commit information again via `git` command, so why we use `prettyLogFormat` parameter? Just parse the full commit information may give us better performance.
Of course this issue is out of this PR. I just found that there.

File: integrations/repo_commits_search_test.go
Comment (gary-kim) [reply integrations/repo_commits_search_test.go:33]: Sorry, I'm not sure what you mean by that.

**Comment by gary-kim** [comment]
>     * If a commit has another commit's id on commit message, we may get duplicated commit output, how to fix that?

I just pushed a commit that should solve this issue.



>     * I find repo. parsePrettyFormatLogToList will get all commit information again via `git` command, so why we use `prettyLogFormat` parameter? Just parse the full commit information may give us better performance.
>       Of course this issue is out of this PR. I just found that there.

I did find that strange as well. Could I open a PR after this to try and address that?

File: integrations/repo_commits_search_test.go
Comment (mrsdizzie) [reply integrations/repo_commits_search_test.go:33]: I think to have a search that should *not* match anything, to test that it works right when there is nothing to match also.

So if there is no commit ```58f73d1a78``` then do a search for that and make sure it returns empty and not something else.

Comment (gary-kim) [reply integrations/repo_commits_search_test.go:33]: Okay, sounds good.

Comment (gary-kim) [reply integrations/repo_commits_search_test.go:33]: Alright, added a few.

**Comment by lunny** [comment]
> > * If a commit has another commit's id on commit message, we may get duplicated commit output, how to fix that?
> 
> I just pushed a commit that should solve this issue.
> 
> > * I find repo. parsePrettyFormatLogToList will get all commit information again via `git` command, so why we use `prettyLogFormat` parameter? Just parse the full commit information may give us better performance.
> >   Of course this issue is out of this PR. I just found that there.
> 
> I did find that strange as well. Could I open a PR after this to try and address that?

Yes, please.

**Review by lunny — APPROVED** [review]

**Comment by gary-kim** [comment]
> I did find that strange as well. Could I open a PR after this to try and address that?

I just took a better look at how `parsePrettyFormatLogToList` works and I think I'll need some time to understand this codebase a little more.

**Comment by typeless** [comment]
@gary-kim There seems to be a more straightforward approach.
1. Scan each keyword and determine if it is a hex number (https://golang.org/pkg/encoding/hex/#DecodeString) and its length is longer than 4 characters (less than 4 characters won't be accepted by git as valid commit id).
2. If it is a hex number, collect it into a slice (named `hashes`, for example).
3.  Append the `hashes` slice to the end of the command.

**Comment by gary-kim** [comment]
> @gary-kim There seems to be a more straightforward approach.
> 
>     1. Scan each keyword and determine if it is a hex number (https://golang.org/pkg/encoding/hex/#DecodeString) and its length is longer than 4 characters (less than 4 characters won't be accepted by git as valid commit id).
> 
>     2. If it is a hex number, collect it into a slice (named `hashes`, for example).
> 
>     3. Append the `hashes` slice to the end of the command.

`git log` only uses the first commit hash in the command so I don't think appending hashes to the end of the command is going to do anything.
Also, there would need to be a check to make sure it is actually a commit first anyways because `parsePrettyFormatLogToList` seems to return an error if any of the given commits are not actual commits.

**Comment by typeless** [comment]
@gary-kim Yeah, that does not work as I expected :(
My rationale is that it would be nice if the seemingly duplicate code paths (`cmd` and `args`) can be unified into one.
It is probably not easy though.

By the way, what is the expected behavior of the hash keyword? Do they OR or AND with other conditions?

**Comment by gary-kim** [comment]
> @gary-kim Yeah, that does not work as I expected :(
> My rationale is that it would be nice if the seemingly duplicate code paths (`cmd` and `args`) can be unified into one.
> It is probably not easy though.
> 
> By the way, what is the expected behavior of the hash keyword? Do they OR or AND with other conditions?

They will act similarly to the keywords. The commit needs to match all the given conditions such as `author:` and at least one of the keywords/hashes.

**Comment by lunny** [comment]
@gary-kim I think it's enough for this PR. @typeless is the original author of this feature. :)

File: modules/git/repo_commit.go
Comment (typeless) [inline modules/git/repo_commit.go]: This should be moved to the `for _, v := range opts.Keywords {` at the bottom of the function.
I think the outcomes will be no different.

**Comment by gary-kim** [comment]
>  @typeless is the original author of this feature. :)

Thank you @typeless for this feature and everyone else who has contributed to Gogs/Gitea.

**Comment by typeless** [comment]
@gary-kim you're welcome. I am a user benefited from this project too.

**Comment by typeless** [comment]
I am not going to block the PR if there is no clear way to improve it. But it still needs another maintainer's review & approval.

**Comment by gary-kim** [comment]
@typeless I got rid of the duplicate code paths. What do you think?

File: modules/git/repo_commit.go
Comment (typeless) [inline modules/git/repo_commit.go]: The constant 6 here is a little bit mysterious.

Comment (gary-kim) [reply modules/git/repo_commit.go]: My bad, I kind of changed that without thinking about it. Changed it back to 4. The idea is it's probably not worth checking strings that are too short.

File: modules/git/repo_commit.go
Comment (gary-kim) [reply modules/git/repo_commit.go:261]: New commits are added on by appending new hashes and a new line but you make a good point. Could just move the new line to before the hash.

EDIT: Done

EDIT2: Turns out that was needed to prevent 500 errors. I put it back in.

Comment (gary-kim) [reply modules/git/repo_commit.go]: New commits are added on by appending new hashes and a new line but you make a good point. Could just move the new line to before the hash.

Comment (typeless) [reply modules/git/repo_commit.go:261]: Okay, it looks like a problem of how parsePrettyFormatLogToList handles its input.

**Comment by typeless** [comment]
@gary-kim LGTM

File: modules/git/repo_commit.go
Comment (typeless) [reply modules/git/repo_commit.go]: I think the constant 4 or 6 deserves a comment. I can figure that it is related to how `git` recognize the shortest allowed input hash. But it is not apparent simply looking at the code.

**Review by typeless — APPROVED** [review]

**Comment by lunny** [comment]
CI failed.

**Comment by gary-kim** [comment]
> CI failed.

Fixed.

**Comment by lunny** [comment]
@typeless you are not a maintainer yet but you have satisfied the condition, please send a PR to apply for that. Otherwise your approval will not be counted.

File: modules/git/repo_commit.go
Comment (guillep2k) [inline modules/git/repo_commit.go:243]: If I'm getting this code right, you always add the first `'\n'` and then always remove the last one with`TrimSuffix()`. Perhaps it would be easier to prepend `'\n'` when required? e.g. change:
`stdout = append(stdout, hashMatching...)`
into
`stdout = append(stdout, '\n', hashMatching...)`
and remove the other appends & TrimSuffix?

**Review by guillep2k — COMMENTED** [review]
Check for simplification on `'\n'` handling.

File: modules/git/repo_commit.go
Comment (gary-kim) [reply modules/git/repo_commit.go:243]: The issue with prepending is that `repo.parsePrettyFormatLogToList` is a very picky function. Having any empty line will result in a 500 error. By prepending, if the first command did not return any results, you could end up with an empty line at the start. You could check for if the line is empty beforehand but then you have an extra if statement that runs very time during the loop.

The important part is that the first `'\n'` is only added if the original command did not return anything.

Also, the multiple appends were needed because Go does not let you use `...` as a part, not the whole of a variadic function parameter.

Comment (guillep2k) [reply modules/git/repo_commit.go:243]: OK, I see it now.

**Review by guillep2k — APPROVED** [review]

**Pull request merged**
